pkg/promtail/targets/syslogparser/syslogparser_test.go

```diff
@@ -0,0 +1,61 @@
+package syslogparser_test
+
+import (
+	"io"
+	"strings"
+	"testing"
+
+	"github.com/grafana/loki/pkg/promtail/targets/syslogparser"
+	"github.com/influxdata/go-syslog/v2"
+	"github.com/stretchr/testify/require"
+)
+
+func TestParseStream_OctetCounting(t *testing.T) {
+	r := strings.NewReader("23 <13>1 - - - - - - First24 <13>1 - - - - - - Second")
+
+	results := make([]*syslog.Result, 0)
+	cb := func(res *syslog.Result) {
+		results = append(results, res)
+	}
+
+	err := syslogparser.ParseStream(r, cb)
+	require.NoError(t, err)
+
+	require.Equal(t, 2, len(results))
+	require.NoError(t, results[0].Error)
+	require.Equal(t, "First", *results[0].Message.Message())
+	require.NoError(t, results[1].Error)
+	require.Equal(t, "Second", *results[1].Message.Message())
+}
+
+func TestParseStream_NewlineSeparated(t *testing.T) {
+	r := strings.NewReader("<13>1 - - - - - - First\n<13>1 - - - - - - Second")
```

Comment:
Actually, I think that this test is wrong. It may seem like a good idea to use that last unfinished line, but the problem is that we don't know if it is a full message or not, as there is no frame separator.  Perhaps message was cut off in the middle? It would be better to log error in such case.

It also brings us back to using `nontransparent.Parser` – it would simplify the code, and do the right thing. Also non-transparent modes could use different trailers (newline, zero byte, or even crlf -- not supported by this lib) -- if we are going to have it, we may need to expose trailer type as an option.

Reply:
Your totally right on the wrong test.

Nontransparent currently supports only newline and null byte. Which we can implement both with scanner and nontransparent. The skipping of the CR happens inside the [rfc5424 parser](https://github.com/influxdata/go-syslog/blob/1193720dd65bc4bab8d9ec1e0c3b5e2496f4d08f/rfc5424/machine.go.rl#L295).

I'm fine with using nontransparent. I just don't like the long, unexpected delays that can happen if there are few messages per connection.

Do you strongly prefer nontransparent? If so I will implement it with nontransparent.
Should I add configuration option for the trailing character with default newline?

Reply:
I am able to reproduce behaviour you're describing (https://gist.github.com/pstibrany/b5fb4e524f478c2f7cd8915aa08f5dc7) but I think it should be reported and fixed in go-syslog package.

Are you running into this bug yourself?

Reply:
> Are you running into this bug yourself?

Yes that's the bug.

Reply:
What I mean was whether it affects your deployment. Could you switch to octet-counting?

After some digging, they actually seem to be doing that to handle multiline messages in syslog, see updated Gist.

Reply:
> What I mean was whether it affects your deployment. Could you switch to octet-counting?

No it does not, we use syslog-ng where the default is octet-counting. Rsyslog on the other hand uses nontransparent by default, but it is easily configurable. Maybe we should include sample configuration for both?

> After some digging, they actually seem to be doing that to handle multiline messages in syslog, see updated Gist.

Yeah looks like it. Most tools I know don't try to handle newlines in syslog. In rsyslog

```
<39>1 - - - - - - First
another line
<39>1 - - - - - - First
```
becomes
```
<39>1 - - - - - - First
<13>1 - another line - - - 
<39>1 - - - - -B - First
```

Reply:
> Maybe we should include sample configuration for both?

That sounds like great idea!

It's funny that rsyslog doesn't handle multiline messages, as it is in [their documentation](https://www.rsyslog.com/doc/v8-stable/configuration/modules/omfwd.html#tcp-framing) where I found this being mentioned as a possibility.

Reply:
:+1:, I will:
- [ ] Use `nontransparent`
- [ ] Document why lines can be delayed.
- [ ] Include sample configuration with octet-counting for syslog-ng and rsyslog under the `syslog_config` section.

Ok?

Reply:
> 👍, I will:
> 
> * [ ]  Use `nontransparent`
> * [ ]  Document why lines can be delayed.
> * [ ]  Include sample configuration with octet-counting for syslog-ng and rsyslog under the `syslog_config` section.
> 
> Ok?

Perfect, thank you!